### PR TITLE
Use Percentages

### PR DESCRIPTION
**Pull request by magopian**
This is a first draft of using "proper" percentages (stored internally as integers) to fix issues like #226.

What do you think @n1k0 , should I continue?

### CODE REVIEW

**Comment by magopian** [comment]
Ok, I believe this is now up for review.

Please @n1k0 could you double check that I didn't mistook "ratios as percentages" (from 0 to 100) and "ratios as floats" (from 0 to 1) in the changes? I tried to be very cautious, and appart from a very minor rounding change in the tests, I didn't spot any deviations in the computation.

I don't think there are other ratios that could be changed to `Split`s, I've searched for `Unit.Ratio` on the codebase, and the only remaining uses are:
- `AggregatedScoreData.weighting`
- `Ingredient.rawToCookedRatio`
- `Preparation.elec`
- `Preparation.heat`

**Comment by magopian** [comment]
@paulboosz if you have automated checks with the gsheet, and would love that you give this a second look too 🙏

File: src/Data/Food/Builder/Recipe.elm [hunk 1]
```diff
@@ -357,13 +358,11 @@ computeIngredientTransport db { ingredient, country, mass, planeTransport } =
         planeRatio =
             -- Special case: if the default origin of an ingredient is "by plane"
             -- and we selected a transport by plane, then we take an air transport ratio of 1
-            Unit.Ratio
-                (if planeTransport == Ingredient.ByPlane then
-                    1
+            if planeTransport == Ingredient.ByPlane then
+                Split.full
```
Comment (n1k0) [inline src/Data/Food/Builder/Recipe.elm:362]: Same, `Spit 100` is more effective to me, matter of taste maybe.

File: src/Data/Food/Builder/Recipe.elm [hunk 2]
```diff
@@ -357,13 +358,11 @@ computeIngredientTransport db { ingredient, country, mass, planeTransport } =
...
 
-                 else
-                    0
-                )
+            else
+                Split.zero
```
Comment (n1k0) [inline src/Data/Food/Builder/Recipe.elm:365]: Same

File: src/Data/Split.elm [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+
+add : Split -> Split -> Result String Split
+add (Split a) (Split b) =
+    let
+        total =
+            a + b
+    in
+    if total < 0 || total > 100 then
+        Err "L'addition de plusieurs part doit être entre comprise entre 0 et 100%"
```
Comment (n1k0) [inline src/Data/Split.elm:122]: s/L'addition/La somme
s/part/parts

File: src/Data/Split.elm [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+
+toFloatString : Split -> String
+toFloatString =
+    asFloat >> String.fromFloat
+
+
+toPercentString : Split -> String
+toPercentString (Split int) =
+    String.fromInt int
```
Comment (n1k0) [inline src/Data/Split.elm:112]: I'm surprise this doesn't contain the unit? (`%`)

File: src/Data/Textile/Inputs.elm [hunk 5]
```diff
@@ -319,15 +320,15 @@ weavingOptionsToString _ _ =
...
     , airTransportRatio
         |> Maybe.andThen
             (\ratio ->
-                if Unit.ratioToFloat ratio == 0 then
+                if Split.asPercent ratio == 0 then
                     Nothing
 
                 else
-                    Just (Format.ratioToPercentString ratio ++ " de transport aérien")
+                    Just (Split.toPercentString ratio ++ " de transport aérien")
```
Comment (n1k0) [inline src/Data/Textile/Inputs.elm:331]: Nit: maybe rename `ratio` to `split` in such circumstances to ease readability a bit

**Review by n1k0 — APPROVED** [review]
Looks good. A bunch of remarks you may want to address or not.

File: src/Data/Split.elm [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+    if int < 0 || int > 100 then
+        Err ("Une part (en pourcentage) doit être comprise entre 0 et 100 inclus (ici: " ++ String.fromInt int ++ ")")
+
+    else
+        Ok (Split int)
+
+
+asFloat : Split -> Float
+asFloat (Split int) =
+    toFloat int / 100
```
[the remark this answers is not in the document]
Comment (magopian) [reply src/Data/Split.elm:97]: yep, I pondered about that too, i'll do it

File: src/Data/Env.elm [hunk 7]
```diff
@@ -37,14 +37,14 @@ githubUrl =
...
 minMakingWasteRatio =
-    Unit.ratio 0
+    Split.zero
 
 
-maxMakingWasteRatio : Unit.Ratio
+maxMakingWasteRatio : Split
 maxMakingWasteRatio =
-    Unit.ratio 0.4
+    Split.fourty
```
[the remark this answers is not in the document]
Comment (magopian) [reply src/Data/Env.elm:47]: For now i'd like to keep the type as Opaque, so `Split.percent 40` would return a Result. All those helpers are for numbers we use in the code, and at least for now there aren't many, so I added them.

**Comment by n1k0** [comment]
Forgot to answer your question; indeed `Preparation.elec` and `Preparation.heat` are percentages and should complete each others.

File: src/Data/Split.elm [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+    asFloat >> String.fromFloat
+
+
+toPercentString : Split -> String
+toPercentString (Split int) =
+    String.fromInt int
+
+
+add : Split -> Split -> Result String Split
+add (Split a) (Split b) =
```
[the remark this answers is not in the document]
Comment (magopian) [reply src/Data/Split.elm:116]: Good idea. When changing that, I realized it's not used anywhere... so actually i'll just drop this code entirely 😬

File: src/Data/Split.elm (hunk 4 above)
Comment (magopian) [reply src/Data/Split.elm:112]: The `%` sign is added in `Format.splitAsPercentage`

File: src/Data/Split.elm [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+    in
+    if total < 0 || total > 100 then
+        Err "L'addition de plusieurs part doit être entre comprise entre 0 et 100%"
+
+    else
+        Ok (Split total)
+
+
+complement : Split -> Split
+complement (Split int) =
```
[the remark this answers is not in the document]
Comment (n1k0) [reply src/Data/Split.elm:129]: I was thinking being able to do `complement [ Split.tenth, Split.twenty ] == Split 70` but now I'm not sure.

File: src/Data/Textile/Inputs.elm (hunk 5 above)
Comment (magopian) [reply src/Data/Textile/Inputs.elm:331]: Not sure about that: the record field name (that corresponds to the API, and the json data) is still `airTransportRatio`.

File: src/Data/Split.elm (hunk 4 above)
Comment (magopian) [reply src/Data/Split.elm:111]: Yep, it's sometimes useful to have a full representation with the % sign and as HTML, sometimes we just need the number "as an int between 0 and 100" without the sign (like in `ShareField` which uses this as the input value)

File: src/Data/Textile/Inputs.elm (hunk 5 above)
Comment (n1k0) [reply src/Data/Textile/Inputs.elm]: I won't die on this hill, please land this patch 😄

**Pull request merged**